bpf/generictracer/protocol_kafka.h

```diff
@@ -0,0 +1,523 @@
+static __always_inline int kafka_send_large_buffer(tcp_req_t *req,
+                                                   pid_connection_info_t *pid_conn,
+                                                   const void *u_buf,
+                                                   u32 bytes_len,
+                                                   u8 direction,
+                                                   enum large_buf_action action) {
+
+    if (kafka_store_state_data(&pid_conn->conn, u_buf, bytes_len, direction) < 0) {
+        bpf_dbg_printk("kafka_send_large_buffer: 4 bytes packet, storing state data");
+        return -1;
+    }
+
+    // check if this response matches an ongoing request
+    struct kafka_correlation_data *correlation_data =
+        bpf_map_lookup_elem(&kafka_ongoing_requests, &pid_conn->conn);
+    if (!correlation_data) {
+        bpf_dbg_printk("kafka_send_large_buffer: no ongoing request found for this response");
+        return 0;
+    }
+    struct kafka_response_hdr hdr = {};
+    if (kafka_read_response_header(u_buf, bytes_len, &hdr) == 0) {
+        bpf_dbg_printk("kafka_send_large_buffer: failed to check kafka response header");
+        return 0;
+    }
+    if (hdr.correlation_id != correlation_data->correlation_id) {
+        bpf_dbg_printk("kafka_send_large_buffer: request correlation_id != response "
+                       "correlation_id, %d != %d",
+                       correlation_data->correlation_id,
+                       hdr.correlation_id);
+        return 0;
+    }
+
+    bpf_map_delete_elem(&kafka_ongoing_requests, &pid_conn->conn);
+
+    tcp_large_buffer_t *large_buf = (tcp_large_buffer_t *)kafka_large_buffers_mem();
+    if (!large_buf) {
+        bpf_dbg_printk("kafka_send_large_buffer: failed to reserve space for Kafka large buffer");
+        return 0;
+    }
+
+    // if we are here it means that the packet is for sure a response because
+    // we populated the hdr with message_size and correlation_id
+    // we checked also in the map of ongoing requests if there was a related
+    // ongoing request so we can hardcode the packet_type as response
+    large_buf->type = EVENT_TCP_LARGE_BUFFER;
+    large_buf->packet_type = PACKET_TYPE_RESPONSE;
+
+    large_buf->action = action;
+    large_buf->direction = direction;
+    large_buf->conn_info = pid_conn->conn;
+    large_buf->tp = req->tp;
+
+    int written = kafka_read_fixup_buffer(
+        &pid_conn->conn, large_buf->buf, &large_buf->len, u_buf, bytes_len, direction);
+    if (written < 0) {
+        bpf_dbg_printk("kafka_send_large_buffer: failed to read buffer, not sending large buffer");
+        return 0;
+    }
+
+    u32 total_size = sizeof(tcp_large_buffer_t);
+    total_size += written > sizeof(void *) ? written : sizeof(void *);
+
+    req->has_large_buffers = true;
+    bpf_ringbuf_output(&events, large_buf, total_size & k_large_buf_max_size_mask, get_flags());
+
+    bpf_dbg_printk(
```

Comment:
how is this working? 🤔

Reply:
what do you mean?

Reply:
oh, you were talking about the `bpf_dbg_printk` I was checking the `bpf_ringbuf_output`. I think it is working because the 4th arg is after 80 characters. I will fix it.